src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs

```diff
@@ -124,107 +122,270 @@ public void Decode(in ReadOnlySequence<byte> data, bool endHeaders, IHttpHeaders
-            DecodeInternal(data, endHeaders, handler);
+            DecodeInternal(data, handler);
             CheckIncompleteHeaderBlock(endHeaders);
         }
 
-        private void DecodeInternal(ReadOnlySpan<byte> data, bool endHeaders, IHttpHeadersHandler? handler)
+        private void DecodeInternal(ReadOnlySpan<byte> data, IHttpHeadersHandler? handler)
         {
-            int intResult;
+            int currentIndex = 0;
 
-            for (int i = 0; i < data.Length; i++)
+            do
             {
-                byte b = data[i];
                 switch (_state)
                 {
                     case State.Ready:
-                        // TODO: Instead of masking and comparing each prefix value,
-                        // consider doing a 16-way switch on the first four bits (which is the max prefix size).
-                        // Look at this once we have more concrete perf data.
-                        if ((b & IndexedHeaderFieldMask) == IndexedHeaderFieldRepresentation)
-                        {
-                            _headersObserved = true;
+                        Parse(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderFieldIndex:
+                        ParseHeaderFieldIndex(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderNameIndex:
+                        ParseHeaderNameIndex(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderNameLength:
+                        ParseHeaderNameLength(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderNameLengthContinue:
+                        ParseHeaderNameLengthContinue(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderName:
+                        ParseHeaderName(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderValueLength:
+                        ParseHeaderValueLength(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderValueLengthContinue:
+                        ParseHeaderValueLengthContinue(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderValue:
+                        ParseHeaderValue(data, ref currentIndex, handler);
+                        break;
+                    case State.DynamicTableSizeUpdate:
+                        ParseDynamicTableSizeUpdate(data, ref currentIndex);
+                        break;
+                    default:
+                        // Can't happen
+                        Debug.Fail("HPACK decoder reach an invalid state");
+                        throw new NotImplementedException(_state.ToString());
+                }
+            }
+            while (currentIndex < data.Length);
```

Comment:
What happens if `data` enters the method as empty?  We no longer check for that before calling the various Parse functions.  Is that impossible?  If so, add an assert at the beginning of the method?

Reply:
The length check is here rather than the top of the method because every parse function checks the length.

I'll add tests that passes in empty data in various states to verify.